lib/internal/modules/esm/loader.js

```diff
@@ -168,100 +212,254 @@ class Loader {
-    }
-    if (getSource !== undefined) {
-      this._getSource = FunctionPrototypeBind(getSource, null);
-    }
-    if (transformSource !== undefined) {
-      this._transformSource = FunctionPrototypeBind(transformSource, null);
-    }
-    if (getGlobalPreloadCode !== undefined) {
-      this._getGlobalPreloadCode =
-        FunctionPrototypeBind(getGlobalPreloadCode, null);
+    const count = specifiers.length;
+    const namespaces = new Array(count);
+
+    for (let i = 0; i < count; i++) {
+      const specifier = specifiers[i];
+      const job = await this.getModuleJob(specifier, parentURL);
+      const { module } = await job.run();
+
+      namespaces[i] = module.getNamespace();
     }
+
+    return wasArr ?
+      namespaces :
+      namespaces[0];
   }
 
-  runGlobalPreloadCode() {
-    if (!this._getGlobalPreloadCode) {
-      return;
+  /**
+   * Provide the ready-to-use source (output is something node understands).
+   *
+   * The internals of this WILL change when chaining is implemented,
+   * depending on the resolution/consensus from #36954
+   * @param {string} url The URL/path of the module to be loaded
+   * @param {Object} context Metadata about the module
+   * @returns {Object}
+   */
+  async load(url, context = {}) {
+    const defaultLoader = this.#loaders[0];
+
+    const loader = this.#loaders.length === 1 ?
+      defaultLoader :
+      this.#loaders[1];
+    const loaded = await loader(url, context, defaultLoader);
+
+    if (typeof loaded !== 'object') {
+      throw new ERR_INVALID_RETURN_VALUE(
+        'object',
+        'loader load',
+        loaded,
+      );
     }
-    const preloadCode = this._getGlobalPreloadCode();
-    if (preloadCode === null) {
-      return;
+
+    const {
+      format,
+      source,
+    } = loaded;
+
+    if (format == null) {
+      const dataUrl = RegExpPrototypeExec(
+        /^data:([^/]+\/[^;,]+)(?:[^,]*?)(;base64)?,/,
+        url,
+      );
+
+      throw new ERR_INVALID_MODULE_SPECIFIER(
+        url,
+        dataUrl ? `has an unsupported MIME type "${dataUrl[1]}"` : ''
+      );
     }
 
-    if (typeof preloadCode !== 'string') {
-      throw new ERR_INVALID_RETURN_VALUE(
-        'string', 'loader getGlobalPreloadCode', preloadCode);
+    if (typeof format !== 'string') {
+      throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
+        'string',
+        'loader resolve',
+        'format',
+        format,
+      );
     }
-    const { compileFunction } = require('vm');
-    const preloadInit = compileFunction(preloadCode, ['getBuiltin'], {
-      filename: '<preload>',
-    });
-    const { NativeModule } = require('internal/bootstrap/loaders');
-
-    preloadInit.call(globalThis, (builtinName) => {
-      if (NativeModule.canBeRequiredByUsers(builtinName)) {
-        return require(builtinName);
+
+    if (
+      source != null &&
+      typeof source !== 'string' &&
+      !(source instanceof SharedArrayBuffer) &&
+      !(source instanceof Uint8Array)
```

Comment:
shouldn't this be any array buffer view?

Reply:
Maybe? I used the ones listed from the [existing ESM docs](https://nodejs.org/api/esm.html#esm_getsource_url_context_defaultgetsource):

> **getSource(url, context, defaultGetSource)**
> …
> Returns: `<Object>`
>   * source `<string>` | `<SharedArrayBuffer>` | `<Uint8Array>`